### ISSUE TITLE
Robust exception handling required for variants and 'bad-data driven tests"

### ISSUE THREAD

**Issue by jeromekelleher** [ga4gh/server#270]
We need custom exceptions for all of the error conditions in the `datamodel/variants.py` file. We also need to test these exceptions, however, since there's no point in making custom exceptions if we don't test them.

One way we might to this is to introduce a new directory in `tests` called `bad-data`, which contains data that is designed to exercise the error cases in our code. We'll need some extra test infrastructure for this --- should we call these `bad-datadriven`, and reuse the datadriven infrastructure?

**Comment by macieksmuga** [comment]
Since different exceptions would presumably be generated by different bad data, such tests would need to be structured differently than the `datadriven` ones so far (i.e., we can't just run all of the exception tests over the totality of the test data - they'd need to be matched: This bad data triggered that exception).

Maybe make a directory of `regression-data` in general, that contains data that will only be run on *certain* specified tests, either ones that are expected to throw an error, or some other outlier behavior that's been problematic in the past?

**Comment by jeromekelleher** [comment]
Yes, that's a good point @macieksmuga --- we will need a different structure. We'd need some sort of map from bad-data file to expected exception. What if we had a directory of files for each exception (that is, any file in that directory is expected to throw a given exception)? Our directory tree of bad-data could mirror the exception hierarchy.

**Comment by jeromekelleher** [comment]
After discussion, it seems that this is a bit ambitious and over complicated. I think the simplest approach is to have a new directory `tests/bad_data/variants` and put some test variant sets in there. These should have descriptive directory names, and contain one (or more) dodgy VCFs.

Then, we have a new test module, `tests/unit/test_bad_data.py`, and we have a bunch of standard unittest.TestCase subclasses. We then have some simple mechanism for associating these tests with some data files (perhaps a class variable containing the list of dodgy data set ids?). For example, we might have something like
```python
class TestVariantSetMetadataMismatch(BadDataTest):
    setIds = ["metadataMismatch1", "metadataMismatch2"]

    def testInstantiation(self):
        for variantSetId in self.setIds:
             path =  self.getFullPath(variantSetId)
             self.assertRaises(
                 exceptions.MetadataMisatchException, variants.HtslibVariantSet, path)
```

**Comment by macieksmuga** [comment]
@jeromekelleher I like it, but I could also see the need for some additional mechanism to pass specific `setIds` to individual tests within a class.

**Comment by dcolligan** [comment]
Fixed in #294
